src/lib/pubkey/pqcrystals/pqcrystals_helpers.h

```diff
@@ -0,0 +1,143 @@
+/*
+ * PQ CRYSTALS Common Helpers
+ *
+ * Further changes
+ * (C) 2024 Jack Lloyd
+ * (C) 2024 René Meusel, Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#ifndef BOTAN_PQ_CRYSTALS_HELPERS_H_
+#define BOTAN_PQ_CRYSTALS_HELPERS_H_
+
+#include <concepts>
+#include <cstdint>
+#include <tuple>
+
+#include <botan/internal/bit_ops.h>
+
+namespace Botan {
+
+// clang-format off
+
+template <std::unsigned_integral T>
+   requires(sizeof(T) <= 4)
+using next_longer_uint_t =
+   std::conditional_t<sizeof(T) == 1, uint16_t,
+   std::conditional_t<sizeof(T) == 2, uint32_t,
+   std::conditional_t<sizeof(T) == 4, uint64_t, void>>>;
+
+template <std::signed_integral T>
+   requires(sizeof(T) <= 4)
+using next_longer_int_t =
+   std::conditional_t<sizeof(T) == 1, int16_t,
+   std::conditional_t<sizeof(T) == 2, int32_t,
+   std::conditional_t<sizeof(T) == 4, int64_t, void>>>;
+
+// clang-format on
+
+template <std::integral T>
+   requires(size_t(sizeof(T)) <= 4)
+constexpr T montgomery_R(T q) {
+   using T_unsigned = std::make_unsigned_t<T>;
+   using T2 = next_longer_uint_t<T_unsigned>;
+   return (T2(1) << (sizeof(T) * 8)) % q;
+}
+
+template <std::integral T>
+   requires(size_t(sizeof(T)) <= 4)
+constexpr T montgomery_R2(T q) {
+   using T2 = next_longer_int_t<T>;
+   return (static_cast<T2>(montgomery_R(q)) * static_cast<T2>(montgomery_R(q))) % q;
+}
+
+template <std::integral T>
+struct eea_result {
+      T gcd;
+      T u;
+      T v;
+};
+
+/**
+ * Run the extended Euclidean algorithm to find the greatest common divisor of a
+ * and b and the Bézout coefficients, u and v.
+ */
+template <std::integral T>
+constexpr eea_result<T> extended_euclidean_algorithm(T a, T b) {
+   if(a > b) {
+      std::swap(a, b);
+   }
+
+   T u1 = 0, v1 = 1, u2 = 1, v2 = 0;
+
+   if(a != b) {
+      while(a != 0) {
+         const T q = b / a;
+         std::tie(a, b) = std::make_tuple(b - q * a, a);
+         std::tie(u1, v1, u2, v2) = std::make_tuple(u2, v2, u1 - q * u2, v1 - q * v2);
+      }
+   }
+
+   return {.gcd = b, .u = u1, .v = v1};
+}
```

Comment:
Perhaps, this exists somewhere else already? Or would be useful somewhere else? Same with the monty-param precalculations above.

Reply:
Yeah, quite surprising if it is not implemented anywhere else. The only EEA I found is within the old mceliece implementation which uses an EEA (`polyn_gf2m::eea_with_coefficients`) on GF elements.

Reply:
I don't think this exists elsewhere currently. Likewise the Montgomery R params, since so far Montgomery over a machine word hasn't been of interest